### PR TITLE
✨ Display warning message for low and out of disk space 

### PR DESCRIPTION
**Pull request by jsaq007**
## What do these changes do?

Display warning message based on user preference for low disk space available.
Displays error message when disk space reaches 0GB

https://z43.fogbugz.com/f/cases/190925
Screenshot 2024-03-05 at 15 56 48

## Related issue/s

- relates to ITISFoundation/osparc-issues#5260
- resolves https://z43.fogbugz.com/f/cases/190925

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

**Review by matusdrobuliak66 — APPROVED** [review]

**Review by GitHK — COMMENTED** [review]
This is already looking nice. As we said in person I would pull the widget in the space where the restart buttons are present.

**Review by ignapas — COMMENTED** [review]
The looks are nice but I don't think it needs to be configured by the user. Also, maybe I would not make the threshold in terms of GB but rather in % of the total.
Imo best would be the user does not have any properties to set and the system warns him after he's consumed 98% of his allocation (or 95%)
The other concern I have is what happens when the threshold gets crossed many times in a short period of time. Could there be a lot of disk activity around the threshold potentially triggering thousands of notifications? If yes, that would be bad

**Comment by jsaq007** [comment]
> The looks are nice but I don't think it needs to be configured by the user. Also, maybe I would not make the threshold in terms of GB but rather in % of the total. Imo best would be the user does not have any properties to set and the system warns him after he's consumed 98% of his allocation (or 95%) The other concern I have is what happens when the threshold gets crossed many times in a short period of time. Could there be a lot of disk activity around the threshold potentially triggering thousands of notifications? If yes, that would be bad

Problem with using % is 5% of 1TB is 50GB would this warrant a low disk warning?
I have not problem not allowing a user to customize the warning level but lest say a user often deal with large file sizes and therefore bring the disk down to 1 or 2 GB maybe they would like to be able to set the threshold to something larger so that they don't get caught out.

**Review by elisabettai — APPROVED** [review]
Very nice!

I like that the flash message persists long enough to allow the user the read the whole message. 

Two minor suggestions:
1. In the flash message, add something like "You can personalize your disk space warning in the User Preferences". In this way users have a hint on how to "disable" the message if it becomes too annoying.
2. In the iframe, where you have the blue box showing "37 GB...", I'd clarify that you refer to disk space. They might it is RAM, for example.

**Review by bisgaard-itis — APPROVED** [review]
Thanks a lot

**Comment by odeimaiz** [comment]
It looks good to me 👌

File: services/static-webserver/client/source/class/osparc/workbench/NodeUI.js [hunk 1]
```diff
@@ -321,6 +333,13 @@ qx.Class.define("osparc.workbench.NodeUI", {
       };
       evaluateLifeCycleIcon();
       this.getNode().addListener("changeVersion", () => evaluateLifeCycleIcon());
+      const indicator = new osparc.workbench.DiskUsageIndicator();
+      indicator.setCurrentNode(node);
+      this.add(indicator, {
+        row: 2,
+        column: 0,
+        colSpan: 4
+      });
```
Comment (odeimaiz) [inline services/static-webserver/client/source/class/osparc/workbench/NodeUI.js:342]: You might want to bring this back to the "usage-indicator" child control

File: services/static-webserver/client/source/class/osparc/workbench/DiskUsageIndicator.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+   https://osparc.io
+
+   Copyright:
+     2020 IT'IS Foundation, https://itis.swiss
+
+   License:
+     MIT: https://opensource.org/licenses/MIT
+
+   Authors:
+     * Odei Maiz (odeimaiz)
```
Comment (odeimaiz) [inline services/static-webserver/client/source/class/osparc/workbench/DiskUsageIndicator.js:14]: Nope...

File: services/static-webserver/client/source/class/osparc/widget/PersistentIframe.js [hunk 3]
```diff
@@ -126,45 +129,51 @@ qx.Class.define("osparc.widget.PersistentIframe", {
       });
       const iframeEl = this._getIframeElement();
       iframeEl.setAttribute("allow", "clipboard-write");
+
+      const buttonContainer = this.__buttonContainer = new qx.ui.container.Composite(new qx.ui.layout.HBox(10).set({
+        alignX: "right",
+        alignY: "middle"
+      }));
```
Comment (odeimaiz) [inline services/static-webserver/client/source/class/osparc/widget/PersistentIframe.js:136]: Thanks!

**Review by pcrespov — COMMENTED** [review]
:+1:

File: services/static-webserver/client/source/class/osparc/dashboard/StudyBrowser.js [hunk 4]
```diff
@@ -541,10 +541,13 @@ qx.Class.define("osparc.dashboard.StudyBrowser", {
         osparc.utils.Utils.setIdToWidget(list, "studiesList");
       }
 
+      /*
+      // Import is disabled until an Export function is implemented
       const importStudyButton = this.__createImportButton();
       const isDisabled = osparc.utils.DisabledPlugins.isImportDisabled();
       importStudyButton.setVisibility(isDisabled ? "excluded" : "visible");
       this._toolbar.add(importStudyButton);
+      */
```
Comment (ignapas) [inline services/static-webserver/client/source/class/osparc/dashboard/StudyBrowser.js:550]: commented code

File: services/static-webserver/client/source/class/osparc/workbench/NodeUI.js [hunk 5]
```diff
@@ -162,6 +162,18 @@ qx.Class.define("osparc.workbench.NodeUI", {
...
+        //     height: 20,
+        //     margin: 4,
+        //     // visibility: "excluded"
+        //   });
+        //   this.add(control, {
+        //     row: 2,
+        //     column: 0,
+        //     colSpan: 3
+        //   });
+        //   break;
```
Comment (ignapas) [inline services/static-webserver/client/source/class/osparc/workbench/NodeUI.js:176]: commented code

File: services/static-webserver/client/source/class/osparc/workbench/DiskUsageController.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      if (data["node_id"] && this.__callbacks[data["node_id"]]) {
+        //  notify
+        this.setDiskUsageNotificationToUI(data);
+        this.__callbacks[data["node_id"]].forEach(cb => {
+          cb(data);
+        })
+      }
+    });
+    this.__callbacks = {};
+    console.log("Executed!!!!")
```
Comment (ignapas) [inline services/static-webserver/client/source/class/osparc/workbench/DiskUsageController.js:44]: console.log

**Review by sanderegg — COMMENTED** [review]
very nice

File: services/static-webserver/client/source/class/osparc/workbench/DiskUsageIndicator.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    this._setLayout(new qx.ui.layout.VBox());
+
+    // Subscribe to disk space threshold - Default 5GB
+    lowDiskSpacePreferencesSettings.addListener("changeLowDiskSpaceThreshold", e => {
+      this.__lowDiskThreshold = e.getData();
+      this.__updateDiskIndicator();
+    }, this);
+  },
+
+  properties: {
```
Comment (ignapas) [inline services/static-webserver/client/source/class/osparc/workbench/DiskUsageIndicator.js:42]: This guy could use at least one more property `usedSpace` (next prs)

Comment (odeimaiz) [reply services/static-webserver/client/source/class/osparc/workbench/DiskUsageIndicator.js:42]: 👍

**Review by odeimaiz — APPROVED** [review]

**Pull request merged**